imod/prepare/topsystem/allocation.py

```diff
@@ -0,0 +1,329 @@
+            return _allocate_cells__at_elevation(top, bottom, head)[0]
+        case ALLOCATION_OPTION.at_first_active:
+            return _allocate_cells__at_first_active(active)[0]
+        case _:
+            raise ValueError(
+                "Received incompatible setting for drains, only"
+                f"'{ALLOCATION_OPTION.at_elevation.name}' and"
+                f"'{ALLOCATION_OPTION.at_first_active.name}' supported."
+                f"got: '{allocation_option.name}'"
+            )
+
+
+def allocate_rch_cells(
+    allocation_option: ALLOCATION_OPTION,
+    active: GridDataArray,
+) -> GridDataArray:
+    match allocation_option:
+        case ALLOCATION_OPTION.at_first_active:
+            return _allocate_cells__at_first_active(active)[0]
+        case _:
+            raise ValueError(
+                "Received incompatible setting for drains, only"
+                f"'{ALLOCATION_OPTION.at_first_active.name}' supported."
+                f"got: '{allocation_option.name}'"
+            )
+
+
+def _is_layered(grid: GridDataArray):
+    return "layer" in grid.sizes and grid.sizes["layer"] > 1
+
+
+@enforced_dim_order
+def _allocate_cells__stage_to_riv_bot(
+    top: GridDataArray,
+    bottom: GridDataArray,
+    stage: GridDataArray,
+    bottom_elevation: GridDataArray,
+) -> tuple[GridDataArray, None]:
+    """
+    Allocate cells inbetween river stage and river bottom_elevation. Compared to
+    iMOD5.6, this is similar to setting IDEFFLAYER=0 in the RUNFILE function.
+
+    Note that ``stage`` and ``bottom_elevation`` are not allowed to have a layer
+    dimension.
+
+    Parameters
+    ----------
+    top: GridDataArray
+        top of model layers
+    bottom: GridDataArray
+        bottom of model layers
+    stage: GridDataArray
+        river stage
+    bottom_elevation: GridDataArray
+        river bottom elevation
+
+    Returns
+    -------
+    GridDataArray
+        River cells
+    """
+    PLANAR_GRID.validate(stage)
+    PLANAR_GRID.validate(bottom_elevation)
+
+    if _is_layered(top):
+        top_layered = top
+    else:
+        top_layered = create_layered_top(bottom, top)
+
+    riv_cells = (stage > bottom) & (bottom_elevation < top_layered)
+
+    return riv_cells, None
+
+
+@enforced_dim_order
+def _allocate_cells__first_active_to_riv_bot(
+    active: GridDataArray,
+    top: GridDataArray,
+    bottom: GridDataArray,
+    bottom_elevation: GridDataArray,
+) -> tuple[GridDataArray, None]:
+    """
+    Allocate cells inbetween first active layer and river bottom elevation.
+    Compared to iMOD5.6, this is similar to setting IDEFFLAYER=-1 in the RUNFILE
+    function.
+
+    Note that ``bottom_elevation`` is not allowed to have a layer dimension.
+
+    Parameters
+    ----------
+    active: GridDataArray
+        active model cells
+    top: GridDataArray
+        top of model layers
+    bottom: GridDataArray
+        bottom of model layers
+    bottom_elevation: GridDataArray
```

Comment:
river_bottom_elevation?

Reply:
bottom_elevation is used in the River package, wanted to stick to that convention as much as possible.